### ISSUE TITLE
Manual calculations gradients are inconsistent with pytorch for LayerNorm

### ISSUE THREAD

**Issue by 0Addicted0** [pytorch/pytorch#141425]
### 🐛 Describe the bug

**Describe the bug**
pytorch's autogradient is inconsistent with the manually calculated gradient for LayerNorm in torch.float32, `atol=1e-5` is required to pass the check, while for torch.float64, this issue does not exist.

**Minimal compelete example**
```python
import torch
import torch.nn as nn

torch.manual_seed(0)
device = "cuda"
dtype = torch.float32
eps = 1e-5
layer = nn.LayerNorm(4, eps=eps, dtype=dtype).to(device)
inp: torch.Tensor = torch.randn((2, 3, 4), dtype=dtype, requires_grad=True).to(device)
inp.retain_grad()
out: torch.Tensor = layer(inp)
out.retain_grad()
criteria = lambda x: 0.5 * x.pow(2).sum()
loss: torch.Tensor = criteria(out)
loss.backward()

with torch.no_grad():
    mean_ = inp.mean(dim=-1, keepdim=True)
    shift = inp - mean_
    var = (shift**2).mean(dim=-1, keepdim=True)
    rstd = torch.rsqrt(var + eps)
    x_hat = shift * rstd
nabla_y = out
hand_weight_grad = (nabla_y * x_hat).sum(dim=(0, 1))
hand_bias_grad = nabla_y.sum(dim=(0, 1))
nabla_x_hat = nabla_y * layer.weight
hand_input_grad = rstd * (
    nabla_x_hat
    - nabla_x_hat.mean(dim=-1, keepdim=True)
    - x_hat * (nabla_x_hat * x_hat).mean(dim=-1, keepdim=True)
)
print("weight.grad check: ", torch.allclose(hand_weight_grad, layer.weight.grad))
print("bias.grad check: ", torch.allclose(hand_bias_grad, layer.bias.grad))
print("input.grad check: ", torch.allclose(hand_input_grad, inp.grad))
# print("input.grad check: ", torch.allclose(hand_input_grad, inp.grad, atol=1e-5))  # float32
```

### Versions

Collecting environment information...
PyTorch version: 2.1.2+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.3) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.18  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-32GB

Nvidia driver version: 535.183.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.1.2+cu118
[pip3] torch-cluster==1.6.3+pt21cu118
[pip3] torch_geometric==2.4.0
[pip3] torch-scatter==2.1.2+pt21cu118
[pip3] torch-sparse==0.6.18+pt21cu118
[pip3] torch-spline-conv==1.2.2+pt21cu118
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.1.2+cu118
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.16.2+cu118
[conda] numpy                     1.24.4           py38h59b608b_0    conda-forge
[conda] torch                     2.1.2+cu118              pypi_0    pypi
[conda] torch-cluster             1.6.3+pt21cu118          pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torch-scatter             2.1.2+pt21cu118          pypi_0    pypi
[conda] torch-sparse              0.6.18+pt21cu118          pypi_0    pypi
[conda] torch-spline-conv         1.2.2+pt21cu118          pypi_0    pypi
[conda] torch-tb-profiler         0.4.3                    pypi_0    pypi
[conda] torchaudio                2.1.2+cu118              pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.16.2+cu118             pypi_0    pypi

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan

**Comment by Paarth-Kashyap** [comment]
Hey I would like to grab this issue if possible. Can you explain a bit more what the problem is?

**Comment by 0Addicted0** [comment]
> Hey I would like to grab this issue if possible. Can you explain a bit more what the problem is?

In torch.float32 precision, I manually calculated the gradient of the layernorm layer, but dx seems to differ from the result of pytorch. However, in torch.float64 precision, the result of manual calculation is consistent with the result of pytorch. I am not sure whether pytorch calculates the gradient internally in certain steps with higher precision or uses a different precision format.

**Comment by Paarth-Kashyap** [comment]
can you help me duplicate this bug? which file is this code snippet from

**Comment by soulitzer** [comment]
Closing for now because it's not expected that a kernel on cuda produces bit-for-bit identical results with manually calling the individual operations.  Comparing the max/mean, relative/absolute errors between autograd float32 and manual float32 with a reference (autograd in float64), it's also not clear to me that the autograd version is less accurate:
```
                  Auto (float32)        Manual (float32)

Input Grad:
  mean abs err:   0.00021333756302107607 0.0002133313831940556
  median abs err: 4.679016541071958e-05 4.674096447583777e-05
  max abs err:    0.0017211254280911194 0.0017219339325012075
  mean rel err:   16.92725940346731 16.905538502596592
  median rel err: 1.8248679439305795 1.8238558703403078
  max rel err:    228.08971457888666 228.1968604830276
```


<details>

<summary>Script to produce differences</summary>

```python
import torch
import torch.nn as nn

# Set the device and dtype
device = "cuda" if torch.cuda.is_available() else "cpu"
eps = 1e-5

def relative_error(x, y, eps=1e-12):
    return torch.abs(x - y)/(torch.abs(y)+eps)

def absolute_error(x, y):
    return torch.abs(x - y)

def max_relative_error(x, y, eps=1e-12):
    return relative_error(x, y, eps).max()

def mean_relative_error(x, y, eps=1e-12):
    return relative_error(x, y, eps).mean()

def median_relative_error(x, y, eps=1e-12):
    return relative_error(x, y, eps).median()

def max_absolute_error(x, y):
    return absolute_error(x, y).max()

def mean_absolute_error(x, y):
    return absolute_error(x, y).mean()

def median_absolute_error(x, y):
    return absolute_error(x, y).median()

###############################################################################
# Utility to run forward/backward in double precision to get reference
###############################################################################
def run_double_reference(eps=1e-5):
    torch.manual_seed(0)
    layer_double = nn.LayerNorm(4, eps=eps, dtype=torch.float64).to(device)
    inp_double = torch.randn((2, 3, 4), dtype=torch.float64, requires_grad=True, device=device)

    # Forward and backward
    out_double = layer_double(inp_double)
    loss_double = 0.5 * out_double.pow(2).sum()
    loss_double.backward()

    ref = {
        'weight_grad': layer_double.weight.grad.detach().clone(),
        'bias_grad': layer_double.bias.grad.detach().clone(),
        'input_grad': inp_double.grad.detach().clone(),
    }

    return ref

###############################################################################
# Run reference in double precision
###############################################################################
ref = run_double_reference(eps=eps)

###############################################################################
# Run float32 autograd
###############################################################################
torch.manual_seed(0)
layer_float32 = nn.LayerNorm(4, eps=eps, dtype=torch.float32).to(device)
inp_float32 = torch.randn((2, 3, 4), dtype=torch.float32, requires_grad=True, device=device)

out_float32 = layer_float32(inp_float32)
loss_float32 = 0.5 * out_float32.pow(2).sum()
loss_float32.backward()

auto_weight_grad = layer_float32.weight.grad.detach().clone().to(torch.float64)
auto_bias_grad = layer_float32.bias.grad.detach().clone().to(torch.float64)
auto_input_grad = inp_float32.grad.detach().clone().to(torch.float64)

###############################################################################
# Run manual float32 computations
###############################################################################
torch.manual_seed(0)
layer_manual = nn.LayerNorm(4, eps=eps, dtype=torch.float32).to(device)
inp_manual = torch.randn((2, 3, 4), dtype=torch.float32, requires_grad=True, device=device)
out_manual = layer_manual(inp_manual)
loss_manual = 0.5 * out_manual.pow(2).sum()
loss_manual.backward()

# Manual gradient computation
with torch.no_grad():
    mean_ = inp_manual.mean(dim=-1, keepdim=True)
    shift = inp_manual - mean_
    var = (shift**2).mean(dim=-1, keepdim=True)
    rstd = torch.rsqrt(var + eps)
    x_hat = shift * rstd

nabla_y = out_manual
hand_weight_grad_float32 = (nabla_y * x_hat).sum(dim=(0, 1))
hand_bias_grad_float32 = nabla_y.sum(dim=(0, 1))
nabla_x_hat = nabla_y * layer_manual.weight
hand_input_grad_float32 = rstd * (
    nabla_x_hat
    - nabla_x_hat.mean(dim=-1, keepdim=True)
    - x_hat * (nabla_x_hat * x_hat).mean(dim=-1, keepdim=True)
)

hand_weight_grad = hand_weight_grad_float32.to(torch.float64)
hand_bias_grad = hand_bias_grad_float32.to(torch.float64)
hand_input_grad = hand_input_grad_float32.to(torch.float64)

###############################################################################
# Compute all errors
###############################################################################
def compute_errors(x, y):
    aerr = absolute_error(x, y)
    rerr = relative_error(x, y)
    return {
        'mean_abs_err':   aerr.mean().item(),
        'median_abs_err': aerr.median().item(),
        'max_abs_err':    aerr.max().item(),
        'mean_rel_err':   rerr.mean().item(),
        'median_rel_err': rerr.median().item(),
        'max_rel_err':    rerr.max().item(),
    }

auto_weight_err = compute_errors(auto_weight_grad, ref['weight_grad'])
auto_bias_err   = compute_errors(auto_bias_grad,  ref['bias_grad'])
auto_input_err  = compute_errors(auto_input_grad, ref['input_grad'])

manual_weight_err = compute_errors(hand_weight_grad, ref['weight_grad'])
manual_bias_err   = compute_errors(hand_bias_grad,   ref['bias_grad'])
manual_input_err  = compute_errors(hand_input_grad,  ref['input_grad'])

###############################################################################
# Print in two-column format
###############################################################################
def print_comparison(grad_type, auto_err, manual_err):
    # Columns: Auto on the left, Manual on the right
    print(f"{grad_type} Grad:")
    print(f"  mean abs err:   {auto_err['mean_abs_err']:<15} {manual_err['mean_abs_err']}")
    print(f"  median abs err: {auto_err['median_abs_err']:<15} {manual_err['median_abs_err']}")
    print(f"  max abs err:    {auto_err['max_abs_err']:<15} {manual_err['max_abs_err']}")
    print(f"  mean rel err:   {auto_err['mean_rel_err']:<15} {manual_err['mean_rel_err']}")
    print(f"  median rel err: {auto_err['median_rel_err']:<15} {manual_err['median_rel_err']}")
    print(f"  max rel err:    {auto_err['max_rel_err']:<15} {manual_err['max_rel_err']}")
    print()

print("Final Gradient Differences vs Float64 Reference:\n")
print("                         Auto (float32)        Manual (float32)\n")
print_comparison("Weight", auto_weight_err, manual_weight_err)
print_comparison("Bias",   auto_bias_err,   manual_bias_err)
print_comparison("Input",  auto_input_err,  manual_input_err)

```

</details>
